### ISSUE TITLE
After update table REPOSITORY cannot be found

### ISSUE THREAD

**Issue by x0r13** [gitbucket/gitbucket#2946]
## Issue
**Impacted version**: 4.37.0

**System**: DEBIAN 4.9.272-2

**Deployment mode**: download gitbucket.war via wget to server

**Problem description**:
After running gitbucket for quite a while without problems, I decided to update today from 4.11 to 4.37. All I did was replacing the existing gitbucket.war file with the new one like I did in the past.
After starting with commandline **java -jar gitbucket.war --host=x.x.x.x --port=8457** an error is thrown:

```
14:00:47.473 [main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
14:00:49.279 [main] ERROR com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Exception during pool initialization.
org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "REPOSITORY" not found; SQL statement:
ALTER TABLE PUBLIC.REPOSITORY ADD CONSTRAINT PUBLIC.IDX_REPOSITORY_PK PRIMARY KEY(USER_NAME, REPOSITORY_NAME) INDEX PUBLIC.PRIMARY_KEY_8E [42102-199]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:451)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:427)
        at org.h2.message.DbException.get(DbException.java:205)
        at org.h2.message.DbException.get(DbException.java:181)
        at org.h2.command.ddl.AlterTableAddConstraint.tryUpdate(AlterTableAddConstraint.java:108)
        at org.h2.command.ddl.AlterTableAddConstraint.update(AlterTableAddConstraint.java:78)
        at org.h2.engine.MetaRecord.execute(MetaRecord.java:60)
        at org.h2.engine.Database.open(Database.java:842)
        at org.h2.engine.Database.openDatabase(Database.java:319)
        at org.h2.engine.Database.<init>(Database.java:313)
        at org.h2.engine.Engine.openSession(Engine.java:69)
        at org.h2.engine.Engine.openSession(Engine.java:201)
        at org.h2.engine.Engine.createSessionAndValidate(Engine.java:178)
        at org.h2.engine.Engine.createSession(Engine.java:161)
        at org.h2.engine.Engine.createSession(Engine.java:31)
        at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:336)
        at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:169)
        at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:148)
        at org.h2.Driver.connect(Driver.java:69)
        ...
```
I checked using h2 (java -cp h2*.jar org.h2.tools.Shell) and 'SHOW TABLES;' does show the REPOSITORY table. Also, all my repositories are shown when using 'SELECT * FROM REPOSITORY;'

However, I'm not able to successfully start the server anymore. Maybe not connected, but before updating gitbucket, I ran a dist upgrade on the Debian system itself. Can anyone please help me to get my data back?

**Comment by takezoe** [comment]
Could show the contents of `VERSIONS` in your H2 database table?

**Comment by h2000** [comment]
I got the same error today as I tried to upgrade from 4.35.3 to 4.37.0 (please see log below).
Versions Table:
```sql
-- 7 +/- SELECT COUNT(*) FROM PUBLIC.VERSIONS;
INSERT INTO "PUBLIC"."VERSIONS" VALUES
('notifications', '1.10.0'),
('gitbucket-core', '4.35.3'),
('h2console', '1.1.1'),
('h2-backup', '1.7.0'),
('gist', '4.20.0'),
('emoji', '4.6.0'),
('pages', '1.9.0');
```

Here the log file snippet:
```log
2021-12-13 23:43:27.332:INFO:oejsh.ContextHandler:JettyShutdownThread: Stopped o.e.j.w.WebAppContext@693e4d19{/gitbucket,null,STOPPED}{file:/home/git/gitbucket-4.35.3.war}
2021-12-13 23:54:17.048:INFO::main: Logging initialized @102ms to org.eclipse.jetty.util.log.StdErrLog
2021-12-13 23:54:17.112:INFO:oejs.Server:main: jetty-9.4.44.v20210927; built: 2021-09-27T23:02:44.612Z; git: 8da83308eeca865e495e53ef315a249d63ba9332; jvm 1.8.0_292-8u292-b10-0ubuntu1~20.04-b10
2021-12-13 23:54:17.955:INFO:oejw.StandardDescriptorProcessor:main: NO JSP Support for /gitbucket, did not find org.eclipse.jetty.jsp.JettyJspServlet
2021-12-13 23:54:18.074:INFO:oejs.session:main: DefaultSessionIdManager workerName=node0
2021-12-13 23:54:18.074:INFO:oejs.session:main: No SessionScavenger set, using defaults
2021-12-13 23:54:18.076:INFO:oejs.session:main: node0 Scavenging every 660000ms
23:54:18.364 [main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
23:54:18.548 [main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Start completed.
23:54:18.559 [main] WARN  slick.util.AsyncExecutor - Having maxConnection > maxThreads can result in deadlocks if transactions or database locks are used.
23:54:18.568 [main] INFO  g.core.servlet.InitializeListener - Check version
23:54:18.568 [main] INFO  g.core.servlet.InitializeListener - Start schema update
23:54:18.989 [main] INFO  l.servicelocator.ServiceLocator - Can not use class liquibase.parser.core.json.JsonChangeLogParser as a Liquibase service because org.yaml.snakeyaml.constructor.BaseConstructor is not in the classpath
23:54:18.990 [main] INFO  l.servicelocator.ServiceLocator - Can not use class liquibase.parser.core.yaml.YamlChangeLogParser as a Liquibase service because org.yaml.snakeyaml.constructor.BaseConstructor is not in the classpath
2021-12-13 23:54:19.442:WARN:oejw.WebAppContext:main: Failed startup of context o.e.j.w.WebAppContext@49aa766b{/gitbucket, ....
org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "REPOSITORY" not found; SQL statement:|ALTER TABLE PUBLIC.REPOSITORY ADD SAFE_MODE BOOLEAN DEFAULT 'true' NOT NULL [42102-199]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:451)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:427)
        at org.h2.message.DbException.get(DbException.java:205)
        at org.h2.message.DbException.get(DbException.java:181)
        at org.h2.command.ddl.AlterTableAlterColumn.update(AlterTableAlterColumn.java:114)
        at org.h2.command.CommandContainer.update(CommandContainer.java:133)
        at org.h2.command.Command.executeUpdate(Command.java:267)
        at org.h2.jdbc.JdbcPreparedStatement.executeUpdateInternal(JdbcPreparedStatement.java:200)
        at org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:154)
        at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
        at io.github.gitbucket.solidbase.migration.MigrationUtils.updateDatabase(MigrationUtils.java:20)
        at io.github.gitbucket.solidbase.migration.LiquibaseMigration.migrate(LiquibaseMigration.java:93)
        at io.github.gitbucket.solidbase.migration.LiquibaseMigration.migrate(LiquibaseMigration.java:54)
        at io.github.gitbucket.solidbase.Solidbase.migrate(Solidbase.java:45)
        at io.github.gitbucket.solidbase.Solidbase.migrate(Solidbase.java:29)
        at gitbucket.core.servlet.InitializeListener.$anonfun$contextInitialized$1(InitializeListener.scala:69)
        at gitbucket.core.servlet.InitializeListener.$anonfun$contextInitialized$1$adapted(InitializeListener.scala:59)
        at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.$anonfun$withTransaction$2(BlockingProfile.scala:207)
        at slick.JdbcProfileBlockingSession$BlockingSession.withTransaction(TransactionalJdbcBackend.scala:26)
        at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.$anonfun$withTransaction$1(BlockingProfile.scala:207)
        at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.withSession(BlockingProfile.scala:200)
        at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.withTransaction(BlockingProfile.scala:207)
        at gitbucket.core.servlet.InitializeListener.contextInitialized(InitializeListener.scala:59)
        at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:1067)
        at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:572)
        at org.eclipse.jetty.server.handler.ContextHandler.contextInitialized(ContextHandler.java:996)
        at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:746)
        at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:379)
        at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1449)
        at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1414)
        at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:910)
        at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:288)
        at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:524)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
        at org.eclipse.jetty.server.handler.StatisticsHandler.doStart(StatisticsHandler.java:253)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:117)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:169)
        at org.eclipse.jetty.server.Server.start(Server.java:423)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:110)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:97)
        at org.eclipse.jetty.server.Server.doStart(Server.java:387)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at JettyLauncher.main(JettyLauncher.java:231)
2021-12-13 23:54:19.450:INFO:oejs.AbstractConnector:main: Started ServerConnector@7d6f77cc{HTTP/1.1, (http/1.1)}{localhost:9981}
2021-12-13 23:54:19.452:INFO:oejs.Server:main: Started @2506ms
2021-12-14 00:02:52.587:INFO:oejs.AbstractConnector:JettyShutdownThread: Stopped ServerConnector@7d6f77cc{HTTP/1.1, (http/1.1)}{localhost:9981}
2021-12-14 00:02:52.587:INFO:oejs.session:JettyShutdownThread: node0 Stopped scavenging
```

**Comment by h2000** [comment]
The repository table before the tried upgrade looked like this: (restored a backup and switched back to 4.35.3, snipped from the h2 script export backup.sql).

```sql
CREATE CACHED TABLE "PUBLIC"."REPOSITORY"(
    "USER_NAME" VARCHAR(100) NOT NULL,
    "REPOSITORY_NAME" VARCHAR(100) NOT NULL,
    "PRIVATE" BOOLEAN NOT NULL,
    "DESCRIPTION" CLOB,
    "DEFAULT_BRANCH" VARCHAR(100),
    "REGISTERED_DATE" TIMESTAMP NOT NULL,
    "UPDATED_DATE" TIMESTAMP NOT NULL,
    "LAST_ACTIVITY_DATE" TIMESTAMP NOT NULL,
    "ORIGIN_USER_NAME" VARCHAR(100),
    "ORIGIN_REPOSITORY_NAME" VARCHAR(100),
    "PARENT_USER_NAME" VARCHAR(100),
    "PARENT_REPOSITORY_NAME" VARCHAR(100),
    "EXTERNAL_ISSUES_URL" VARCHAR(200),
    "EXTERNAL_WIKI_URL" VARCHAR(200),
    "ALLOW_FORK" BOOLEAN DEFAULT TRUE NOT NULL,
    "WIKI_OPTION" VARCHAR(10) DEFAULT 'DISABLE' NOT NULL,
    "ISSUES_OPTION" VARCHAR(10) DEFAULT 'DISABLE' NOT NULL,
    "MERGE_OPTIONS" VARCHAR(200) DEFAULT 'merge-commit,squash,rebase' NOT NULL,
    "DEFAULT_MERGE_OPTION" VARCHAR(100) DEFAULT 'merge-commit' NOT NULL
);
ALTER TABLE "PUBLIC"."REPOSITORY" ADD CONSTRAINT "PUBLIC"."IDX_REPOSITORY_PK" PRIMARY KEY("USER_NAME", "REPOSITORY_NAME");
-- 6 +/- SELECT COUNT(*) FROM PUBLIC.REPOSITORY;
INSERT INTO "PUBLIC"."REPOSITORY" VALUES (...)
```

**Comment by takezoe** [comment]
I guess H2 was broken if your database could be recovered by the recovery tool. I strongly recommend taking a backup before upgrading and rather using an external database (MySQL or PostgreSQL) instead of the default embedded H2 if would use use GitBucket for important purposes.

By the way, 4.37.1 could solve the notification plugin error:
https://github.com/gitbucket/gitbucket/releases/tag/4.37.1

**Comment by h2000** [comment]
@takezoe thanks a lot for your comments and the nice peace of software :)

**Comment by h2000** [comment]
Using the backup data and running gitbucket in the old version, I exported the database data to sql and followed: https://github.com/gitbucket/gitbucket/wiki/External-database-configuration#data-migration to switch from H2 to postgres.
After this I could update gitbucket to 4.37.1 :)

**Comment by h2000** [comment]
@x0r13 : Would it be okay for you if I close the ticket?

**Comment by x0r13** [comment]
Since I was able to run the current version with my (restored) backup file, yes I think it can be closed. Thanks.

**Comment by Fmstrat** [comment]
@SIkebe I don't think this should be closed. OP had to go through a restore process to get this working, but it's still an issue. Starting up in Docker using `gitbucket/gitbucket:4.35.3` and running fine, but then updating to `gitbucket/gitbucket:4.36.1` or `latest` results in this error. There must be a missing step in the migration.

**Comment by tomk3003** [comment]
I had the same issue today upgrading from 4.31.2 to 4.37.2.
I really would like to be able to update without switching to another db engine.

**Comment by takezoe** [comment]
I don't think it's not an auto migration issue because it's a set of normal SQL operations via JDBC which is impossible to break database itself. For the maximum safety, It's would be better to conduct upgrading version by version in an isolated environment from client requests.

Anyway, we strongly recommend to not use the default embedded H2 database for production use. H2 database could be broken when JVM crashes, etc even without upgrading.

**Comment by mdhub** [comment]
Hi *,
just for your info.
1. I did "Data export" from my 4.35.3 instance.
2. I started a new empty 4.36.0 instance on a different host to create a new H2 database (data.mv.db).
3. I imported my database export from 1.) into the new empty database (gitbucket version 4.36.0)
4. I stopped the  gitbucket version 4.36.0 instance and copied the database data.mv.db to my old 4.35.3 instance.
5. Now i could start my old 4.35.3 instance with the 'new' gitbucket.war file (version 4.36.0) without errors.

best regards,
Michael
